evm/src/cpu/kernel/tests/bn254.rs

```diff
@@ -0,0 +1,188 @@
+use std::str::FromStr;
+
+use anyhow::Result;
+use ethereum_types::U256;
+
+use crate::bn254::{
+    curve_generator, fp12_to_vec, frob_fp12, gen_fp12, gen_fp12_sparse, miller_loop, mul_fp12,
+    power, tate, twisted_curve_generator, Curve, Fp12, TwistedCurve,
+};
+use crate::cpu::kernel::aggregator::KERNEL;
+use crate::cpu::kernel::interpreter::run_interpreter;
+
+fn make_label(lbl: &str) -> U256 {
+    U256::from(KERNEL.global_labels[lbl])
+}
+
+fn make_stack(vecs: Vec<Vec<U256>>) -> Vec<U256> {
+    let mut stack = vec![];
+    for vec in vecs {
+        stack.extend(vec)
+    }
+    stack
+}
+
+fn get_output(lbl: &str, stack: Vec<U256>) -> Vec<U256> {
+    let label = KERNEL.global_labels[lbl];
+    let mut input = stack;
+    input.reverse();
+    let mut output = run_interpreter(label, input).unwrap().stack().to_vec();
+    output.reverse();
+    output
+}
+
+fn make_mul_stack(f: Fp12, g: Fp12, mul_label: &str) -> Vec<U256> {
+    let in0 = U256::from(64);
+    let in1 = U256::from(76);
+    let out = U256::from(88);
+
+    make_stack(vec![
+        vec![in0],
+        fp12_to_vec(f),
+        vec![in1],
+        fp12_to_vec(g),
+        vec![
+            make_label(mul_label),
+            in0,
+            in1,
+            out,
+            make_label("return_fp12_on_stack"),
+            out,
+        ],
+    ])
+}
+
+#[test]
```

Comment:
These integration tests are sufficient when things are working. But if something breaks, whoever has to fix it will have a lot of trouble working out the actual function that's broken. It would be nice to test the smaller functions too, e.g. `test_frob_fp6`, `test_mul_fp2`, etc.